language/move-binary-format/src/compatibility.rs

```diff
@@ -268,3 +268,89 @@ fn type_parameter_phantom_decl_compatibile(
     // old_type_paramter.is_phantom => new_type_parameter.is_phantom
     !old_type_parameter.is_phantom || new_type_parameter.is_phantom
 }
+
+/// A simpler, and stricter compatibility checker relating to the inclusion of the old module in
+/// the new.
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq)]
+pub enum InclusionCheck {
+    Subset,
+    Equal,
+}
+
+impl InclusionCheck {
+    pub fn check(&self, old_module: &Module, new_module: &Module) -> PartialVMResult<()> {
+        let err = Err(PartialVMError::new(
+            StatusCode::BACKWARD_INCOMPATIBLE_MODULE_UPDATE,
+        ));
+
+        // Module checks
+        if old_module.address != new_module.address
+            || old_module.name != new_module.name
+            || old_module.file_format_version != new_module.file_format_version
+        {
+            return err;
+        }
+
+        // If we're checking exactness we make sure there's an inclusion, and that the size of all
+        // of the tables are the exact same.
+        if (self == &Self::Equal)
+            && (old_module.structs.len() != new_module.structs.len()
+                || old_module.exposed_functions.len() != new_module.exposed_functions.len()
+                || old_module.private_functions.len() != new_module.private_functions.len()
+                || old_module.friends.len() != new_module.friends.len()
+                || old_module.constants.len() != new_module.constants.len())
+        {
+            return err;
+        }
+
+        // Struct checks
+        for (name, old_struct) in &old_module.structs {
+            match new_module.structs.get(name) {
+                Some(new_struct) if old_struct == new_struct => (),
+                _ => {
+                    return err;
+                }
+            };
+        }
+
+        // Function checks
+        for (name, old_func) in old_module
+            .exposed_functions
+            .iter()
+            .chain(old_module.private_functions.iter())
+        {
+            match new_module
+                .exposed_functions
+                .get(name)
+                .or_else(|| new_module.private_functions.get(name))
+            {
+                Some(new_func) if old_func == new_func => (),
+                _ => {
+                    return err;
+                }
+            }
+        }
+
+        // Friend checks
+        let new_friends: BTreeSet<_> = new_module.friends.iter().collect();
+        if !old_module
+            .friends
+            .iter()
+            .all(|old_friend| new_friends.contains(old_friend))
+        {
+            return err;
+        }
+
+        // Constant checks
+        let new_constants: BTreeSet<_> = new_module.constants.iter().collect();
```

Comment:
By a similar argument to the above for friendships, I'm not strictly sure we need to check the constant pool.  (If a constant gets added to the pool but nothing uses it, did the module change?)

Reply:
For the subset case agreed. But for the `Equal` check, I think _technically_ it's fine to not check this, but it feels kinda strange to me to not check it. I think I'll also remove these along with the friendship checks to relax things a bit, and we can tighten them down later if need be.